Test gen

rjust opened the pull request:
This PR adds a generic `gen_tests` script, which calls any of the supported test generators through a simple interface: it exports a set of variables and calls a wrapper shell script which invokes the test generator.

This PR includes wrapper scripts for evosuite and randoop and a test script, which Travis executes in the first testing stage.

This PR also adds a new parameter for test generation, test mode: regression or error-revealing.

jose commented on framework/bin/gen_tests.pl:
```diff
@@ -0,0 +1,284 @@
+# Export all environment variables that are expected by the wrapper script of
+# the test generator.
+$ENV{D4J_HOME}                = "$BASE_DIR";
+$ENV{D4J_FILE_ALL_CLASSES}    = "$LOADED_CLASSES";
```
You might want to rename this variable to `D4J_FILE_LOADED_CLASSES` as it does not list **all** classes.

jose commented on framework/bin/gen_tests.pl:
```diff
@@ -0,0 +1,284 @@
+# the test generator.
+$ENV{D4J_HOME}                = "$BASE_DIR";
+$ENV{D4J_FILE_ALL_CLASSES}    = "$LOADED_CLASSES";
+$ENV{D4J_FILE_TARGET_CLASSES} = "$MOD_CLASSES";
```
You might want to rename this variable to `D4J_FILE_MOD_CLASSES` or `D4J_FILE_MODIFIED_CLASSES`. I find the name `_TARGET_` in this context quite confusing.

jose commented on framework/bin/gen_tests.pl:
```diff
@@ -0,0 +1,284 @@
+$ENV{D4J_DIR_TESTGEN_LIB}     = "$TESTGEN_LIB_DIR";
+$ENV{D4J_CLASS_BUDGET}        = "$TIME";
+# Use test_id and bug_id to compute the random seed!
+$ENV{D4J_SEED}                = ($TID*1000 + $BID);
```
This is more a personal request than a comment: could you please make `D4J_SEED` an optional parameter? I usually like to have full control of that parameter. :smile:

Greg4cr replied:
I agree with this.

jose commented on framework/lib/test_generation/generation/bin/evosuite.sh:
```diff
@@ -0,0 +1,50 @@
+# General helper functions
+source $D4J_DIR_TESTGEN_LIB/bin/_tool.source
+
+for class in $(cat $D4J_FILE_TARGET_CLASSES); do
```
The `framework/bin/gen_tests.pl` script says that:

> Tests are generated for all classes touched by at least one triggering test.

however in here EvoSuite is only invoked for modified classes... As I commented in the `framework/bin/gen_tests.pl` script, I think the `gen_tests` module should (by default) generate test cases for all modified classes or allow people to provide a custom list of classes.

Greg4cr replied:
Yes. By default, I would have test generators target only the modified classes, and allow a custom list as an option. Most of the test gen literature on D4J targets only modified classes to control cost.

mernst replied:
It's true that is done when running Evosuite, but I find it artificial and unrealistic.  A programmer wants tests for their entire program, not just part of it (and doesn't know the part relevant to one particular bug).

rjust replied:
Targeted test generation does make sense for regression testing (fixed version); I agree that it is not realistic when generating error-revealing tests (buggy version).

I updated the script to generate tests for the modified classes by default (consistent with the prior implementation), but a user can provide a set of classes that the test generator should target.

What is currently missing is the notion of there are two different use cases: generating regression tests vs. generating error-revealing tests. I suggest that we tackle this once this general test-gen interface is merged.

rjust replied:
I added the two test modes.

jose commented on framework/bin/gen_tests.pl:
```diff
@@ -0,0 +1,284 @@
+=head1 DESCRIPTION
+
+This script runs the specified test generator on a particular program version.
+Tests are generated for all classes touched by at least one triggering test.
```
People might want to generate test cases for different sets of classes. Could you please make this an optional parameter? I would say that the script should by default generate test cases for all modified classes but it should also allow people to provide a custom list of classes.

Greg4cr replied:
Agreed

jose commented on framework/lib/test_generation/generation/bin/randoop.sh:
```diff
@@ -0,0 +1,72 @@
+          --only-test-public-members=false \
+          --omitmethods=HashCodeAndEqualsSafeSet.of \
+          --regression-test-basename=$BASE_NAME \
+          --require-covered-classes=$D4J_FILE_TARGET_CLASSES"
```
(same comment I added to EvoSuite's properties)

jose commented on framework/lib/test_generation/generation/bin/evosuite.sh:
```diff
@@ -0,0 +1,50 @@
+    -Djunit_check=false \
+    -Dfilter_assertions=false \
+    -Dtest_comments=false \
+    -mem 1500"
```
How about we move these properties to a `evosuite.config` file? I think it would be easier for people to have N configuration files than having to edit this particular script every time they want to generate cases with a different EvoSuite configuration.

Greg4cr replied:
Agreed

Greg4cr commented:
I agree with @jose's comments and I would suggest a four-file system for all test generators:
- <generator>.sh: Wrapper script needed to call the file
- <generator>.credit: Credit file
- <generator>.config: Properties/Command-line parameters.
- <generator>.default.config: Default command-line parameters. 

Both EvoSuite and Randoop has several command-line parameters that may be altered for test generation. This should be handled in a way that does not require changing the core script. I definitely run experiments where I feed in all sorts of different EvoSuite configurations, and it would be risky to touch the "evosuite.sh" file with automated scripts. 

Instead, scripts could edit the <generator>.config file. The <generator>.default.config file is there to retain the "default" values, and can be used to restore <generator>.config back to "normal".

Greg4cr commented:
Addressing the core questions in your opening post:
**Are the default configurations in the wrapper scripts suitable?**

They seem fine, but I would also put them in a separate config file (see above)

**Are the "credit" files correct?**

They are the papers I would choose. You can, of course, contact the authors and ask what they would prefer.

**Should this PR immediately delete the obsolete scripts run_evosuite.pl and run_randoop.pl?**

Yes. Some experiment scripting (including my own) will need to be updated, but I think it would be confusing to have multiple ways to run test generation in the framework.

**Should this PR immediately delete the randoop/evosuite functionality from Project.pm?**

See above. Let's enable test generation in a consistent, universal manner.

rjust commented:
Hi @Greg4cr, @jose, @mernst,

I addressed all the comments. Can you please take another look and let me know whether this is ready for merge?

Note that I also added support for error-revealing vs. regression test generation. 

Thanks,
René

Greg4cr commented:
I have added a script that can test the test generation process for one or more specified bugs: framework/test/test_generate_suites.sh. I think this branch is ready to merge into master once we decide on which we want to do of the following:
1) Keep test_generate_suites.sh and test_test_gen.sh seperate.
2) Replace test_test_gen.sh with test_generate_suites.sh (I don't care what filename we use).

rjust commented:
Thanks for providing the script, Greg!

To avoid confusion, I removed the old test_gen_tests.sh script and updated the travis config.
